util/src/migration/tests.rs

```
@@ -14,19 +14,50 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-use std::collections::BTreeMap;
-use migration::{Error, Destination, Migration, Manager, Config};
+//! Tests for migrations.
+//! A random temp directory is created. A database is created within it, and migrations
+//! are performed in temp sub-directories.
 
-impl Destination for BTreeMap<Vec<u8>, Vec<u8>> {
-	fn commit(&mut self, batch: BTreeMap<Vec<u8>, Vec<u8>>) -> Result<(), Error> {
-		self.extend(batch);
-		Ok(())
+use common::*;
+use migration::{Config, SimpleMigration, Manager};
+use kvdb::{Database, DBTransaction};
+
+use devtools::RandomTempPath;
+use std::path::PathBuf;
+
+fn db_path(path: &Path) -> PathBuf {
+	let mut p = path.to_owned();
+	p.push("db");
+	p
+}
+
+// initialize a database at the given directory with the given values.
+fn make_db(path: &Path, pairs: BTreeMap<Vec<u8>, Vec<u8>>) {
+	let db = Database::open_default(path.to_str().unwrap()).expect("failed to open temp database");
+	{
+		let transaction = DBTransaction::new();
+		for (k, v) in pairs {
+			transaction.put(&k, &v).expect("failed to add pair to transaction");
+		}
+
+		db.write(transaction).expect("failed to write db transaction");
+	}
+}
+
+// helper for verifying a migrated database.
+fn verify_migration(path: &Path, pairs: BTreeMap<Vec<u8>, Vec<u8>>) {
+	let db = Database::open_default(path.to_str().unwrap()).unwrap();
+
+	for (k, v) in pairs {
+		let x = db.get(&k).unwrap().unwrap();
```

test code -- fail hard and fail fast